### PR TITLE
Alex's branch

### PR DESCRIPTION
**Pull request by apapiez**
# New Features
---
- Created FactionController which attaches to the GameMaster
- Created Faction class to hold data on Factions
- Faction data is manipulated using the FactionController (something like ```gameMaster.factionController.getFactionPower("Nobility");```

### CODE REVIEW

File: Assets/Scripts/Faction.cs [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+using System.Collections.Generic;
+using UnityEngine;
+
+[System.Serializable]
+public class Faction
+{
+    /// <summary>
+    /// name of the faction
+    /// </summary>
+    [SerializeField] private string factionName; //make this field adjustable in the inspector
```
Comment (apapiez) [inline Assets/Scripts/Faction.cs:11]: Move this down to the other public string FactionName on LINE 19 CAUSE IT'S CONFUSING HERE BEfore the comment

File: Assets/Scripts/Faction.cs [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+        }
+        //if the amount is negative, subtract the amount from the faction happiness and clamp it to a min of 0
+        else
+        {
+            factionHappiness += amount;
+            if (factionHappiness < 0)
+            {
+                factionHappiness = 0;
+            }
+        }
```
Comment (apapiez) [inline Assets/Scripts/Faction.cs:95]: Add a happiness switch statement to check over it.

Possibly add it to the faction controller as a 
Check Happiness Events()
 switch(happiness)
   if(20 to 40)
     etc

File: Assets/Scripts/Faction.cs [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+        //if the amount is negative, subtract the amount from the faction happiness with the AI and clamp it to a min of 0
+        else
+        {
+            factionAiHappiness += amount;
+            if (factionAiHappiness < 0)
+            {
+                factionAiHappiness = 0;
+            }
+        }
+    }
```
Comment (Thalpy) [inline Assets/Scripts/Faction.cs:129]: Same as above

File: Assets/Scripts/Faction.cs [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+            if (factionPower + amount > 100)
+            {
+                factionPower = 100;
+            }
+            else
+            {
+                factionPower += amount;
+            }
+        }
+    }
```
Comment (Thalpy) [inline Assets/Scripts/Faction.cs:163]: Consider an event here too. 
Do we want them to cease to be at 0%?

File: Assets/Scripts/FactionController.cs [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+    public Faction SelectFaction(string factionName)
+    {
+        for (int i = 0; i < factions.Count; i++)
+        {
+            if (factions[i].FactionName == factionName)
+            {
+                return factions[i];
+            }
+        }
+        return null;
```
Comment (Thalpy) [inline Assets/Scripts/FactionController.cs:96]: Add a warning/breakpoint here if we fail to find one.
Probably a breakpoint cause it's serious if we fail.

Comment (Thalpy) [reply Assets/Scripts/FactionController.cs:96]: Unless we're removing factions when they hit 0.

File: Assets/Scripts/FactionController.cs [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+    public float GetFactionPower(string factionName)
+    {
+        for (int i = 0; i < factions.Count; i++)
+        {
+            if (factions[i].FactionName == factionName)
+            {
+                return factions[i].FactionPower;
+            }
+        }
+        return 0;
```
Comment (Thalpy) [inline Assets/Scripts/FactionController.cs:117]: Similar to above, consider adding a more alarming failure state system.

File: Assets/Scripts/Faction.cs (hunk 1 above)
Comment (apapiez) [reply Assets/Scripts/Faction.cs:11]: Done

File: Assets/Scripts/FactionController.cs (hunk 5 above)
Comment (apapiez) [reply Assets/Scripts/FactionController.cs:96]: log and breakpoint added.

File: Assets/Scripts/Faction.cs (hunk 4 above)
Comment (apapiez) [reply Assets/Scripts/Faction.cs:163]: will write function on Faction Controller

File: Assets/Scripts/Faction.cs (hunk 3 above)
Comment (apapiez) [reply Assets/Scripts/Faction.cs:129]: discussed works as is

File: Assets/Scripts/Faction.cs (hunk 2 above)
Comment (apapiez) [reply Assets/Scripts/Faction.cs:95]: Discussed - will add a switch case to the FactionController

File: Assets/Scripts/FactionController.cs (hunk 6 above)
Comment (apapiez) [reply Assets/Scripts/FactionController.cs:117]: Done

File: Assets/Scripts/Faction.cs [hunk 7]
```diff
@@ -93,6 +129,36 @@ public void ChangeHappiness(float amount)
...
+        else if (factionHappiness > 25)
+        {
+            //if the faction happiness is already neutral, do nothing
+            if (playerHappiness == PlayerHappinessEnum.neutral)
+            {
+                return;
+            }
+            PlayerHappiness = PlayerHappinessEnum.neutral;
+        }
+        else
```
Comment (Thalpy) [inline Assets/Scripts/Faction.cs:152]: For if trees switch statements are generally preferred. But this is fine here

File: Assets/Scripts/FactionController.cs (hunk 5 above)
Comment (Thalpy) [reply Assets/Scripts/FactionController.cs:98]: Nice

**Review by Thalpy — APPROVED** [review]
All good!

**Pull request merged**